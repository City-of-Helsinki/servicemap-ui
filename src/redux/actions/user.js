import LocalStorageUtility from '../../utils/localStorage';
import fetchAddress from '../../views/MapView/utils/fetchAddress';

export const setLocale = locale => ({
  type: 'SET_LOCALE',
  locale,
});

export const setInitialLoad = () => ({
  type: 'SET_INITIAL_LOAD',
});

export const setPage = page => ({
  type: 'SET_CURRENT_PAGE',
  page,
});

export const setUserPosition = position => ({
  type: 'SET_POSITION',
  position,
});

const setTheme = theme => ({
  type: 'SET_THEME',
  theme,
});


export const changeLocaleAction = locale => async (dispatch) => {
  dispatch(setLocale(locale));
};

export const actionSetInitialLoad = () => async (dispatch) => {
  dispatch(setInitialLoad());
};

export const setCurrentPage = page => async (dispatch) => {
  dispatch(setPage(page));
};

export const changeTheme = theme => async (dispatch) => {
  LocalStorageUtility.saveItem('theme', theme);
  dispatch(setTheme(theme));
};

export const findUserLocation = () => async (dispatch) => {
  const success = (position) => {
    if (position.coords.accuracy < 10000) {
      fetchAddress({ lat: position.coords.latitude, lng: position.coords.longitude })
        .then((data) => {
          dispatch(setUserPosition({
            coordinates: position.coords,
            allowed: true,
            addressData: data,
          }));
        });
    } else {
      console.warn(`Position accuracy: ${position.coords.accuracy}. Max accuracy: 10000`);
      dispatch(setUserPosition({ coordinates: null, allowed: true, addressData: null }));
    }
  };

  const error = (err) => {
    console.warn(`GeoLocation error:(${err.code}): ${err.message}`);
    if (err.code === 1) {
      dispatch(setUserPosition({ coordinates: null, allowed: false, addressData: null }));
    } else {
      dispatch(setUserPosition({ coordinates: null, allowed: true, addressData: null }));
    }
  };

  navigator.geolocation.getCurrentPosition(success, error);
};


export default { setLocale, changeLocaleAction };
