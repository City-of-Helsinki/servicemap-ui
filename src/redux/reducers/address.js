const initialState = {
  units: [],
  addressCoordinates: null,
  addressData: null,
  adminDistricts: [],
  toRender: 'adminDistricts',
};

export default (state = initialState, action) => {
  switch (action.type) {
    case 'SET_ADDRESS_DATA':
      return {
        ...state,
        addressData: action.data,
      };
    case 'SET_ADDRESS_LOCATION':
      return {
        ...state,
        addressCoordinates: action.location.addressCoordinates,
      };
    case 'SET_ADDRESS_UNITS':
      return {
        ...state,
        units: action.units,
      };
    case 'SET_ADMINISTRATIVE_DISTRICTS':
      return {
        ...state,
        adminDistricts: action.data,
      };
    case 'SET_TO_RENDER':
      return {
        ...state,
        toRender: action.data,
      };
    default:
      return state;
  }
};
