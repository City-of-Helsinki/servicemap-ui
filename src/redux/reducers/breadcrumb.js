const initialState = [];

const breadcrumbReducer = (state = initialState, action) => {
  switch (action.type) {
    case 'PUSH_ENTRY':
      return [...state, action.entry];
    case 'POP_ENTRY': {
      const newEntries = Array.from(state);
      newEntries.pop();
      return newEntries;
    }
    case 'REPLACE_ENTRY': {
      const newEntries = Array.from(state);
      newEntries.pop();
      return [...newEntries, action.entry];
    }
    default:
      return state;
  }
};

export default breadcrumbReducer;
