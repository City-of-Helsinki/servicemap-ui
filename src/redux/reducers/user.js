
const initialState = {
  initialLoad: false,
  locale: 'fi',
  page: 'home',
  theme: 'default',
  position: {
    coordinates: null,
    allowed: false,
  },
};

export default (state = initialState, action) => {
  switch (action.type) {
    case 'SET_LOCALE':
      return {
        ...state,
        locale: action.locale,
      };
    case 'SET_INITIAL_LOAD':
      return {
        ...state,
        initialLoad: true,
      };
    case 'SET_CURRENT_PAGE':
      return {
        ...state,
        page: action.page,
      };
    case 'SET_POSITION':
      return {
        ...state,
        position: action.position,
      };
    case 'SET_THEME':
      return {
        ...state,
        theme: action.theme,
      };
    default:
      return state;
  }
};
