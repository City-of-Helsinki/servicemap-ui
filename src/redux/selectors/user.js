
export const getTheme = state => state.user.theme;
export const getPage = state => state.user.page;
export const getLocale = state => state.user.locale;
