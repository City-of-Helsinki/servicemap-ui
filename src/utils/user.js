import { useSelector } from 'react-redux'
 
export const useUserTheme = () => useSelector(state => state.user.theme);
export const useUserLocale = () => useSelector(state => state.user.locale);
