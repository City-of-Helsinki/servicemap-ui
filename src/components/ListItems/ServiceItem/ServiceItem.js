import React from 'react';
import PropTypes from 'prop-types';
import { drawServiceIcon } from '../../../views/MapView/utils/drawIcon';
import { uppercaseFirst } from '../../../utils';
import SimpleListItem from '../SimpleListItem';

class ServiceItem extends React.Component {
  state = {
    icon: <img alt="" src={null} style={{ height: 24 }} aria-hidden="true" />,
  };

  componentDidMount() {
    const { classes } = this.props;
    this.setState({ icon: <img alt="" src={drawServiceIcon()} className={classes.icon} aria-hidden="true" /> });
  }

  render() {
    const {
      currentService, service, getLocaleText, navigator, setNewCurrentService, link,
    } = this.props;
    const { icon } = this.state;
    let text = getLocaleText(service.name);

    if (service.clarification) {
      text += `: ${getLocaleText(service.clarification)}`;
    }

    const onClick = link ? (e) => {
      e.preventDefault();

      if (!currentService || currentService.id !== service.id) {
        setNewCurrentService(service);
      }
      if (navigator) {
        navigator.push('service', service.id);
      }
    } : null;

    return (
      <SimpleListItem
        button={link}
        text={uppercaseFirst(text)}
        icon={icon}
        divider
        handleItemClick={onClick}
      />
    );
  }
}

export default ServiceItem;

ServiceItem.propTypes = {
  classes: PropTypes.objectOf(PropTypes.any).isRequired,
  currentService: PropTypes.objectOf(PropTypes.any),
  getLocaleText: PropTypes.func.isRequired,
  navigator: PropTypes.objectOf(PropTypes.any),
  service: PropTypes.objectOf(PropTypes.any).isRequired,
  setNewCurrentService: PropTypes.func.isRequired,
  link: PropTypes.bool,
};

ServiceItem.defaultProps = {
  currentService: null,
  navigator: null,
  link: true,
};
