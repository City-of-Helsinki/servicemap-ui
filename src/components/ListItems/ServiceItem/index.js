import { connect } from 'react-redux';
import { withStyles } from '@mui/styles';
import ServiceItem from './ServiceItem';
import styles from './styles';

// Listen to redux state
const mapStateToProps = (state) => {
  const { navigator } = state;
  return {
    navigator,
  };
};

export default withStyles(styles)(connect(
  mapStateToProps,
)(ServiceItem));
