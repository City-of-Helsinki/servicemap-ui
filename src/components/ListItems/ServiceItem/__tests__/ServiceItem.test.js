import React from 'react';
import { initialState } from '../../../../redux/reducers/user';
import ServiceItem from '../index';
import { getRenderWithProviders } from '../../../../../jestUtils';

const mockData = {
  "name": {
      "fi": "yleiset kirjastot",
      "sv": "allm√§nna bibliotek",
      "en": "public libraries"
  },
  "id": 813,
  "clarification_enabled": false,
};

// Generic required props for SimpleListItem
const mockProps = {
  service: mockData,
};

const renderWithProviders = getRenderWithProviders({
  user: initialState,
  settings: {},
  service: {
    current: null,
  },
})

describe('<ServiceItem />', () => {
  it('should work', () => {
    const { container } = renderWithProviders(<ServiceItem {...mockProps} />);
    expect(container).toMatchSnapshot();
  });

  it('does render accessibility attributes correctly', () => {
    const { container } = renderWithProviders(<ServiceItem {...mockProps} />);
    const items = container.querySelectorAll('li');
    const firstItem = items[0];
    const firstItemResultTitle = firstItem.querySelectorAll('p')[0];
    const dividerItem = items[1];

    // List item's image should be aria-hidden
    expect(firstItem.querySelector('img').getAttribute('aria-hidden')).toBeTruthy();
    expect(firstItem.getAttribute('role')).toEqual('link');
    expect(firstItem.getAttribute('tabIndex')).toEqual('0');
    // expect(firstItemSRText.className.indexOf('ResultItem-title') > 0).toBeTruthy();
    // expect(firstItemResultTitle.className.indexOf('ResultItem-title') > 0).toBeTruthy();
    expect(firstItemResultTitle.getAttribute('aria-hidden')).toBeFalsy();
    expect(dividerItem.getAttribute('aria-hidden')).toBeTruthy();
  });
});
