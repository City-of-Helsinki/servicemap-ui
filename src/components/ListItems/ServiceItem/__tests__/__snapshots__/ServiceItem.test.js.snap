// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ServiceItem /> should work 1`] = `
<div>
  <li
    class="MuiListItem-root SimpleListItem-listItem-2 MuiListItem-gutters MuiListItem-padding null  css-sn7obt-MuiListItem-root"
    tabindex="-1"
  >
    <div
      aria-hidden="true"
      class="MuiListItemIcon-root SimpleListItem-listIcon-6  css-cveggr-MuiListItemIcon-root"
    >
      <img
        alt=""
        aria-hidden="true"
        class="Connect(ServiceItem)-icon-1"
        src="test-file-stub"
        style=""
      />
    </div>
    <div
      class="MuiListItemText-root SimpleListItem-textContainer-3 css-vt7cuc-MuiListItemText-root"
    >
      <p
        class="MuiTypography-root null  MuiTypography-body2 css-q91igz-MuiTypography-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body2 css-1ak20dk-MuiTypography-root"
          style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
        />
        Yleiset kirjastot
      </p>
    </div>
  </li>
  <li
    aria-hidden="true"
  >
    <hr
      class="MuiDivider-root MuiDivider-fullWidth SimpleListItem-divider-7 css-9mgopn-MuiDivider-root"
    />
  </li>
</div>
`;
